### DISCUSSION TITLE
Model Viewer for my web app?

### DISCUSSION THREAD

**Discussion by matt998759** [google/model-viewer#4549] [General]
Good evening,

I woke up this morning to news that our 3D/AR provider is going bankrupt and will be shutting it's servers down. We used them for modeling and hosting our ~200 product assets.

Background:
I run an ecommerce site that allows customers to personalize items. They can preview item with their customizations in 3D/AR via browser. 

When customers request 3D, we are generating JPG/PNG design files and passing those to our provider viewer to swap out the texture.

I was able to pull my GLB files from them before the servers went down.  I also have USDZ and GLTF files for what it's worth but I believe the GLTF will have external references to provider servers so these are probably useless.

I have been looking a lot into model-viewer today and am hopeful this could be a solution.  Based on my very brief explanation would model-viewer be a sufficient replacement?

I attached an example GLB file I have just for reference. [16oz Insulated Tumbler (Clasp Top) (1).zip](https://github.com/google/model-viewer/files/13194221/16oz.Insulated.Tumbler.Clasp.Top.1.zip)

Anyone that has knowledge on this your help is greatly appreciated.  I would be willing to pay someone for consulting hours.

Thank you,

Matt

**Comment by zpennachi** [comment]
Yes Matt, I think modelviewer is exactly what you need to replace your previous 3D provider. (image attached is your model displaying in modelviewer)
![image](https://github.com/google/model-viewer/assets/95557128/d641c14d-cc75-4976-a8ee-86da28f561bd)

Not sure of your background but if you need someone to build the intended functionality, let me know!

**Reply by matt998759** [reply]
Thank you! I will be in touch.

**Comment by WadeMorrison** [comment]
Hey there @matt998759! I think it depends on how much functionality/granularity you are wanting with your products.  If you are wanting simple color/size changes, I would say MV is good. They don't have great options for hiding/showing parts, but you can always "hack" your way down to ThreeJS and do it there. 

If you're wanting full control I would say just use ThreeJS, and then you can always export your model to MV (onto a different page or something) to get the AR functionality. I'm still relatively new to programming, so what I'm listing below is still a work in progress. But will hopefully show what can be done with MV! I have built my app from no experience at all using MV, so hopefully that shows how easy it is to set up. Documentation is great, and I feel the community is helpful here as well.

[Simple chair with just color swapping](https://fathomless-brook-11892.herokuapp.com/product/64e6566b3ec4190ccdecb5ba)
[A night stand that changes size, color, and hides/shows handles (WIP)](https://fathomless-brook-11892.herokuapp.com/product/64d3d6a7a9c14146065f4c95)

**Reply by matt998759** [reply]
Really cool, thanks for example.  Will let you know how it goes but we are going to try and get this done in the next week.

**Reply by elalish** [reply]
We actually just made a simple example showing how to hide/show parts using our MV API: https://modelviewer.dev/examples/scenegraph/#hideShowMeshVariantsExample

**Reply by elalish** [reply]
Also, if anyone has a cuter model to contribute to this example, PRs welcome! As you may have guessed, I'm not much of a 3D artist...

**Comment by matt998759** [comment]
@zpennachi @WadeMorrison Our GLB files have several materials and textures defined by a UUID (see attached screenshot).  We are trying to dynamically replace a material by UUID.  How can we get that UUID from modelViewerTexture.model.materials?

![image](https://github.com/google/model-viewer/assets/131872692/5c19e81f-5d49-4e23-b40d-5fb28df59359)

**Reply by WadeMorrison** [reply]
We actually used sayduck for a couple months two years ago. Interesting to see someone else who had also used them. 

As far as I know, MV doesn't support this out of the box, the way I got around it is actually using their [texture swaping code here](https://modelviewer.dev/examples/scenegraph/#swapTexturesExample). And if you look in one of my [examples](https://fathomless-brook-11892.herokuapp.com/product/64e6566b3ec4190ccdecb5ba) the swatch has different data attributes that point to different maps, and once clicked, it runs the texture swap function. 

Which also, @elalish I noticed the other day when I swap my textures and then go into AR, the model shows up as a ghost/see through model instead of textured. On IOS, auto-generated, not sure about Android. It used to work just fine, I tried updating to the current MV version but no luck. Wasn't sure if this would be a Safari thing or the auto-generated feature?

**Reply by matt998759** [reply]
Thanks @WadeMorrison , yea SayDuck surprised us by filing bankruptcy with zero notice and the servers will go offline any day now.  

Thanks for the example.  I think texture swap should work, however we have several materials/textures in the GLB and I am not quite sure how we can specify the texture ID that we need to swap.  As you can see in the screenshot I posted, the UUID is in the "name" which is also not a constant (not always called Artwork).  Essentially, we need to go through each material and find the texture UUID that we need, then swap it.

If you have any ideas on that it would be greatly appreciated.

**Reply by ronpadz** [reply]
@matt998759. So sorry to hear that SayDuck is leaving you in a lurch. We may be able to help if you're finding that building your own solution is too much to manage in a short amount of time. Contact me at: info at nvzn dot net.

**Reply by WadeMorrison** [reply]
If I remember correctly, the UUID is pulled/referenced from an external material, it's not actually inside the GLB (they may have changed the way they do it since I last used it). So you had an external material library basically. I haven't seen any examples on here or from other people who have done it like that. If the material is inside the GLB, your best bet is to use the [materialVariant API](https://modelviewer.dev/examples/scenegraph/#variants), otherwise I would use the texture swap method if you're wanting more of an "external library" of sorts. 

I'll download your model and take a look at the inside of the GLB model and see what I can come up with if you'd like. No charge, as I will be doing this alongside my own job (so wont be as quick as a dedicated person), and it'll be good to test my app with someone else's model. 

Is your background in 3D modeling, coding, or something else?

**Reply by matt998759** [reply]
I really appreciate it @WadeMorrison .  I should have put a better GLB example so I am attaching a new one here.  This is a phone case.  

For this example, we would be dynamically replacing the texture with the UUID: 71aca9f0-c697-013b-7d52-2e500ea4e81a

You will see that is here: 

![image](https://github.com/google/model-viewer/assets/131872692/37f80322-894e-4ca7-9187-2d2c07096602)

As for my background, I am not a developer.  I am more of a solution architect that would bring this to our dev for implementation.  Currently, this is the roadblock they are experiencing.

I appreciate any assistance on this.  Thank you!

[71a32170-c697-013b-7bfa-2e500ea4e81a (2).zip](https://github.com/google/model-viewer/files/13296853/71a32170-c697-013b-7bfa-2e500ea4e81a.2.zip)

**Reply by WadeMorrison** [reply]
Okay, I will look into them both and see what I can come up with.

Do you have discord or an email I can keep in contact with you, so we don't keep blowing up the discussion board?

**Reply by WadeMorrison** [reply]
Hey Matt, I sent you an email, not sure if you got it or not. But here is a really basic example:
https://glitch.com/edit/#!/lake-crawling-hoof

**Comment by matt998759** [comment]
@zpennachi any ideas on this?  I am happy to pay you or anyone that has solution to that a consulting fee.  We are obviously in a bit of a time crunch and this is where we are getting hung up.  Any help would be greatly appreciated. Thank you!

**Reply by zpennachi** [reply]
hey Matt! I actually don't have much experience with UUIDs, it sounds like Wade is gonna be a much better help for you! I'm here to offer free support as well, but I think Wades gonna be your dude! 

If it helps at all, I am currently in the process of building a [CMS-driven material for modelviewer.](https://www.zpennachi.com/product-template/webflow-07091) This is using the textureswapping method that @WadeMorrison mentioned above.


CODE: 

(html controls for selecting image)
<div class="controls">
      <select id="normals2">
        <option>None</option>
        <option value="DYNAMIC CMS CONTENT 1">building</option>
        <option value="DYNAMIC CMS CONTENT 2">learning</option>
        <option value="DYNAMIC CMS CONTENT 3">having fun :)</option>
      </select>
      <p id="image-name">None</p>
      <p id="texture-name">None</p>
    </div>

<script type="module">
const modelViewerTexture = document.querySelector("model-viewer#duck");
const textureName = document.querySelector('#texture-name');
const imageName = document.querySelector('#image-name');

modelViewerTexture.addEventListener("load", () => {

  const material = modelViewerTexture.model.materials[0];

  const createAndApplyTexture = async (channel, event) => {
    if (event.target.value == "None") {
      // Clears the texture.
      material[channel].setTexture(null);
      // Display the names values
      textureName.innerText = "None";
      imageName.innerText = "None";
    } else if (event.target.value) {
      // Creates a new texture.
      const texture = await modelViewerTexture.createTexture(event.target.value);
      // Set the texture name
      texture.name = event.target.options[event.target.selectedIndex].text.replace(/ /g, "_").toLowerCase();
      // Applies the new texture to the specified channel.
      material[channel].setTexture(texture);
      // Display the names values
      textureName.innerText = texture.name;
      imageName.innerText = texture.source.name;
    }
  }

  document.querySelector('#normals2').addEventListener('input', (event) => {
    createAndApplyTexture('normalTexture', event);
  });
});

</script>

**Comment by matt998759** [comment]
Yes, please email me at mattannen at gmail dot com.
On 11/8/2023 8:22:05 AM, WadeMorrison ***@***.***> wrote:
Okay, I will look into them both and see what I can come up with.
Do you have discord or an email I can keep in contact with you, so we don't keep blowing up the discussion board?
—
Reply to this email directly, view it on GitHub [https://github.com/google/model-viewer/discussions/4549#discussioncomment-7510651], or unsubscribe [https://github.com/notifications/unsubscribe-auth/A7ODPNEMB6N3LUDMUFYCAATYDOBP3AVCNFSM6AAAAAA6TTTEMSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TKMJQGY2TC].
You are receiving this because you were mentioned.Message ID: ***@***.***>
[241a4be2-ef5b-4c9c-8d4f-dc55f48dc72e]

**Comment by matt998759** [comment]
@WadeMorrison thank you for your help, we are back up and running and have replaced Sayduck.  Their bankruptcy was a blessing I guess since this will save us about $8,000 a year using model-viewer in our environment with no view or model limitations.  Everything is running like it was before.  Thank you all for your help.  Let me know if you want a free phone case or something thru our site.  Least I can do.

**Comment by sugarfreejaz** [comment]
@WadeMorrison  Looking for more work?

**Reply by WadeMorrison** [reply]
Sure, depending on the scope of it! Email me at 
wademorrison at ymail.com
Or
wade at scoutdesignstudio.com

let me know what you’ve got!

**Reply by sugarfreejaz** [reply]
email sent
